From d4285d3bd051577aa54032f6075bd7224545a9a9 Mon Sep 17 00:00:00 2001
From: rkvenc forward-port <forward-port@local>
Date: Wed, 22 Jul 2026 08:13:33 -0700
Subject: [PATCH 64/87] video: rockchip: rga: clean shutdown jobs outside irq
 lock

Detach queued jobs while holding scheduler->irq_lock, then perform
IOMMU teardown and request signaling after interrupts are restored.
This avoids sleeping cleanup under a spinlock.

Signed-off-by: rkvenc forward-port <forward-port@local>
---
 drivers/video/rockchip/rga3/rga_job.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/drivers/video/rockchip/rga3/rga_job.c b/drivers/video/rockchip/rga3/rga_job.c
index 2bd6bfe2617c6..ecc2b26bd6b6d 100644
--- a/drivers/video/rockchip/rga3/rga_job.c
+++ b/drivers/video/rockchip/rga3/rga_job.c
@@ -756,6 +756,7 @@ void rga_request_scheduler_shutdown(struct rga_scheduler_t *scheduler)
 {
 	struct rga_job *job, *job_q;
 	unsigned long flags;
+	LIST_HEAD(list_to_free);
 
 	rga_power_enable(scheduler);
 
@@ -787,15 +788,21 @@ void rga_request_scheduler_shutdown(struct rga_scheduler_t *scheduler)
 		 */
 		rga_power_disable(scheduler);
 	} else {
-		/* Clean up the jobs in the todo list that need to be free. */
+		/* Move the todo jobs that need to be freed to a local list. */
 		list_for_each_entry_safe(job, job_q, &scheduler->todo_list, head) {
+			list_move(&job->head, &list_to_free);
+			scheduler->job_count--;
+		}
+
+		spin_unlock_irqrestore(&scheduler->irq_lock, flags);
+
+		/* Clean up outside the lock since the callees may sleep. */
+		list_for_each_entry_safe(job, job_q, &list_to_free, head) {
 			rga_mm_unmap_job_info(job);
 
 			job->ret = -EBUSY;
 			rga_request_release_signal(scheduler, job);
 		}
-
-		spin_unlock_irqrestore(&scheduler->irq_lock, flags);
 	}
 
 finish:
